File: px-checkout/src/main/java/com/mercadopago/android/px/model/PaymentRecovery.java
package com.mercadopago.android.px.model;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import java.io.Serializable;

//TODO move to internal package.
public class PaymentRecovery implements Serializable {

    private final String statusDetail;
    private final Token token;
    private final Card card;
    private final PaymentMethod paymentMethod;

    @Deprecated
    public PaymentRecovery(final String paymentStatusDetail) {
        this(paymentStatusDetail, null, null);
    }

    @Deprecated
    public PaymentRecovery(@NonNull final String statusDetail, @NonNull final Token token) {
        this(statusDetail, token, null);
    }

    @Deprecated
    public PaymentRecovery(@NonNull final String statusDetail, @NonNull final Token token, @Nullable final Card card) {
        this(statusDetail, token, card, null);
    }

    public PaymentRecovery(@NonNull final String statusDetail, @NonNull final Token token, @Nullable final Card card,
        @NonNull final PaymentMethod paymentMethod) {
        this.statusDetail = statusDetail;
        this.token = token;
        this.card = card;
        this.paymentMethod = paymentMethod;
    }

    public Token getToken() {
        return token;
    }

    @Nullable
    public Card getCard() {
        return card;
    }

    @NonNull
    public PaymentMethod getPaymentMethod() {
        return paymentMethod;
    }

    public boolean isTokenRecoverable() {
        return Payment.StatusDetail.isStatusDetailRecoverable(statusDetail);
    }

    public boolean isStatusDetailCallForAuthorize() {
        return Payment.StatusDetail.STATUS_DETAIL_CC_REJECTED_CALL_FOR_AUTHORIZE.equals(statusDetail);
    }

    public boolean isStatusDetailCardDisabled() {
        return Payment.StatusDetail.STATUS_DETAIL_CC_REJECTED_CARD_DISABLED.equals(statusDetail);
    }

    public boolean isStatusDetailInvalidESC() {
        return Payment.StatusDetail.STATUS_DETAIL_INVALID_ESC.equals(statusDetail);
    }

    public boolean shouldAskForCvv() {
        return !Payment.StatusDetail.STATUS_DETAIL_CC_REJECTED_CARD_DISABLED.equals(statusDetail);
    }
}

File: px-checkout/src/main/java/com/mercadopago/android/px/model/PaymentRecovery.kt
package com.mercadopago.android.px.model

import android.os.Parcel
import com.mercadopago.android.px.internal.util.KParcelable
import com.mercadopago.android.px.internal.util.parcelableCreator
import java.io.Serializable

//TODO move to internal package.
class PaymentRecovery(private val statusDetail: String, val token: Token?, val card: Card?,
    val paymentMethod: PaymentMethod?) : Serializable, KParcelable {

    @Deprecated("")
    constructor(paymentStatusDetail: String) : this(paymentStatusDetail, null, null)

    @Deprecated("")
    constructor(statusDetail: String, token: Token) : this(statusDetail, token, null)

    @Deprecated("")
    constructor(statusDetail: String, token: Token?, card: Card?) : this(statusDetail, token, card, null)

    val isTokenRecoverable = Payment.StatusDetail.isStatusDetailRecoverable(statusDetail)

    val isStatusDetailCallForAuthorize = Payment.StatusDetail.STATUS_DETAIL_CC_REJECTED_CALL_FOR_AUTHORIZE == statusDetail

    val isStatusDetailCardDisabled = Payment.StatusDetail.STATUS_DETAIL_CC_REJECTED_CARD_DISABLED == statusDetail

    val isStatusDetailInvalidESC = Payment.StatusDetail.STATUS_DETAIL_INVALID_ESC == statusDetail

    constructor(parcel: Parcel) : this(
        parcel.readString()!!,
        parcel.readSerializable() as Token,
        parcel.readParcelable(Card::class.java.classLoader),
        parcel.readParcelable(PaymentMethod::class.java.classLoader))

    fun shouldAskForCvv(): Boolean {
        return Payment.StatusDetail.STATUS_DETAIL_CC_REJECTED_CARD_DISABLED != statusDetail
    }

    override fun writeToParcel(parcel: Parcel, flags: Int) {
        parcel.writeString(statusDetail)
        parcel.writeSerializable(token)
        parcel.writeParcelable(card, flags)
        parcel.writeParcelable(paymentMethod, flags)
    }

    companion object {
        @JvmField
        val CREATOR = parcelableCreator(::PaymentRecovery)
    }
}